analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs

```diff
@@ -50,9 +50,10 @@ private static SymbolicConstraint LearnBranchingConstraint(ProgramState state, I
         {
             return pattern.WrappedOperation.Kind switch
             {
-                OperationKindEx.ConstantPattern => ConstraintFromConstantPattern(state, As(IConstantPatternOperationWrapper.FromOperation), useOpposite),
+                OperationKindEx.ConstantPattern => ConstraintFromConstantPattern(state, As(IConstantPatternOperationWrapper.FromOperation), useOpposite, pattern.InputType.IsReferenceType),
```

Comment:
I would move all the "IsReferenceType" checks to the `public LearnBranchingConstraint`. At the moment we always return null if IsReferenceType is false, but the check is e.g. missing for `OperationKindEx.TypePattern`. We could short-circuit if we know that there is nothing to learn in advance.

Reply:
That wouldn't play nicely with constant and negated constant patterns, as there we don't need reference types to learn bool constraints